### PR TITLE
fix(lua-protobuf): enable encode_default_values to avoid breaking behavior

### PR DESCRIPTION
**Pull request by vm-001**
### Checklist

- [ ] The Pull Request has tests
- [ ] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

**Review by sabertobihwy — APPROVED** [review]

**Review by samugi — APPROVED** [review]

**Comment by StarlightIbuki** [comment]
Could it be the other way around, that we should interpret omitted values as default values?

**Comment by StarlightIbuki** [comment]
Could it be the other way around, that we should interpret omitted values as default values? I seem to remember we have had an issue related to the default value emitting. @javierguerragiraldez

**Comment by StarlightIbuki** [comment]
But it might be good so we have no ambiguity and headaches handling omitted values.

**Comment by javierguerragiraldez** [comment]
defaults are _very_ hard, and this being a global setting means that any change will definitely affect all uses.  perfect recipe for disaster.

the [standard](https://protobuf.dev/programming-guides/proto3/#default) specifies that for scalar values, missing elements are to be interpreted as the default value.  It also specifies that "there’s **no** way of telling whether a field was explicitly set to the default value (for example whether a boolean was set to false) or just not set at all"

> ...once a message is parsed there’s no way of telling whether a field was explicitly set to the default value [] or just not set at all
> For example, don’t have a boolean that switches on some behavior when set to false if you don’t want that behavior to also happen by default.
> Also note that if a scalar message field is set to its default, the value will not be serialized on the wire.

From this, we could say that:

1. the `'encode_default_values'` option is non-standard.

2. any "fix" that depends on default valued being present or not covers an input requirement.  That makes for a very fragile protocol.

... but all this is only for scalar values.  "For message fields, the field is not set. Its exact value is language-dependent."  I guess that's because on static languages, the "empty" value of a static structure has empty values on all it's fields, while on dynamic languages, the "empty" structure is an empty hash table.

for wRPC, i had to do a lot of post-decoding massaging because the config functions depend on the presence of empty tables for all the main sections, and the JSON configurations reflect that, but the protobuf encoding didn't.

In this case, it's not about our protocol, but on the user's, so we can't preach on "how should the receiving end behave".  Perhaps it requires all values to be present... perhaps it does care about the difference between default and missing.

Personally, I think unconditionally adding them is just as dangerous as removing them.  If the option is between always adding the values and always stripping them, I would first try with always stripping, since that's the published specification.

The only defensible deviation from standard would be to reflect the incoming data, and including them if they're present in the input.  I don't know if the current version allows for that.

**Comment by vm-001** [comment]
@javierguerragiraldez 
Another way is to change that broken test case of the internal repository, which may break the different teams. And I'm also concerned that this may also break the users too if they are relaying the previous library behavior.

**Comment by hanshuebner** [comment]
Please make a decision, remove the `pr/do not merge` label and merge.

**Comment by StarlightIbuki** [comment]
| Option                  | Description                                                  |
| ----------------------- | ------------------------------------------------------------ |
|...|...|
| `auto_default_values`   | act as `use_default_values` for proto3 and act as `no_default_values` for the others **(default)** |
| `no_default_values`     | do not default values for decoded message table |
| `use_default_values`    | set default values by copy values from default table before decode |
|...|...|

From official documentation.


And here's a result of search:

```search
9 results - 3 files

kong/tools/grpc.lua:
  42  local function set_hooks()
  43:   pb.option("enable_hooks")
  44  

spec/03-plugins/37-opentelemetry/01-otlp_spec.lua:
  48      -- overwrite for testing
  49:     pb.option("enum_as_value")
  50:     pb.option("auto_default_values")
  51    end)

  54      -- revert it back
  55:     pb.option("enum_as_name")
  56:     pb.option("no_default_values")
  57    end)

spec/03-plugins/37-opentelemetry/04-exporter_spec.lua:
  36        -- overwrite for testing
  37:       pb.option("enum_as_value")
  38:       pb.option("auto_default_values")
  39      end)

  42        -- revert it back
  43:       pb.option("enum_as_name")
  44:       pb.option("no_default_values")
  45      end)
```

My guess is the `pb.option` in tests makes them flaky and dependent on the order of tests. However, we cannot have multiple copies of pb state, and we can neither store the original setting to restore it after (we are assuming `no_default_values` is the previous state and blindly set it).

The short-term solution could be to fix the specific test, using the same save and restore method.

**Comment by vm-001** [comment]
I did a quick test and looks like the encoded bytes also changed.

```lua
--  The example proto on PR description

--pb.option("encode_default_values")
local bytes = pb.encode("RequestMetadata", payload)
print("bytes after pb.encode() using resty.string#to_hex：", str.to_hex(bytes))
print("bytes after pb.encode() using pb.tohex: ", pb.tohex(bytes))

local decoded = pb.decode("RequestMetadata", bytes)
print("decoded table from encoded bytes", require("inspect")(decoded))
```

**Output of lua-protobuf 0.4.2**

```
$ resty test.lua

bytes after pb.encode() using resty.string#to_hex：
bytes after pb.encode() using pb.tohex: 
decoded table from encoded bytes{
  client_ip = "",
  started_at = 0
}
```

**Output of lua-protobuf 0.3.3**

```
$ resty test.lua

bytes after pb.encode() using resty.string#to_hex：1a00
bytes after pb.encode() using pb.tohex: 1A 00
decoded table from encoded bytes{
  client_ip = "",
  route = {
    id = "",
    name = ""
  },
  started_at = 0
}
```

**Output of lua-protobuf 0.4.2 with pb.option("encode_default_values")**

```
$ lua test.lua 

bytes after pb.encode() using resty.string#to_hex：1a040a001200
bytes after pb.encode() using pb.tohex: 1A 04 0A 00 12 00
decoded table from encoded bytes{
  client_ip = "",
  route = {
    id = "",
    name = ""
  },
  started_at = 0
}
```

**Review by ms2008 — COMMENTED** [review]
Cc @zhongchen for visibility. This could impact the analytics module that Konnect used as well.

**Comment by vm-001** [comment]
Add to `3.3.0` milestone as the EE has a dirty fix during bump up lua-protobuf(https://github.com/Kong/kong-ee/pull/4698/commits/a0aed772f8069fd8bb928425ec6275ef8fdec024), and need to be reverted and waiting this one merged.

**Comment by kikito** [comment]
@vm-001 Should this still have the "Do not merge" label? If we don't start working on it soon, it will lose the 3.3 train

**Comment by kikito** [comment]
@vm-001 does this PR belong to any ticket?

I have read this carefully and I think I agree with @javierguerragiraldez .

The TLDR of this is: the lua-protobuf change from 0.3.3 to 0.4.2 does not look like a breaking change, but like a *fix* (the library was doing something in 0.3.3 which is not part of the standard, that is, always adding default values to parameters, which is non-standard behavior).

The config option  `encode_default_values` seems to be there as "backwards compatibility layer for people relying on that non-standard behavior". Setting it "globally" (as it's being done here, or in https://github.com/Kong/kong-ee/commit/a0aed772f8069fd8bb928425ec6275ef8fdec024) would be a valid "quickfix" if we were (the Kong developers) where the only users of the library and needed the non-standard behavior.

However, we are not the only users. We have a couple plugins out (gRPC-gateway and gRPC-web) which other people use. Using the global setting (or keeping using 0.3.3) will give them non-standard behavior, which we should avoid if possible.

I see two options:
* If we (Kong) don't need the default values explicitly set on all the messages, we should remove them (close this PR without merging, and also revert the quick hack in https://github.com/Kong/kong-ee/commit/a0aed772f8069fd8bb928425ec6275ef8fdec024 )
* If we (Kong) do need the default values set on some messages, then we should refrain from using the global setting, and set the default values when needed in "spot places". It seems the `pb.defaults()` function might be used to "fill in the default values" in this fashion (I am unclear about this, the docs for this function are not very good).

@javierguerragiraldez , could you validate my evaluation?

**Comment by javierguerragiraldez** [comment]
> Using the global setting (or keeping using 0.3.3) will give them non-standard behavior, which we should avoid if possible.

full agree on this.   since it's a global setting and the effects are visible to users, the closer it is to the standard, the better.

if going closer to the standard than it was possible in previous versions breaks something, then that something is buggy and should be fixed.

as mentioned, for wRPC i had to do post-decoding massaging because the config loader is fragile in that sense.  fortunately, that code is gone now, so it's one less variable.

**Comment by vm-001** [comment]
@kikito This one doesn't belong to any ticket. The library's behavior change was caught by EE test case.

I created EE PR(https://github.com/Kong/kong-ee/pull/5276) for taking the first option, and also ping other team to evaluate. 

Thanks. @kikito @javierguerragiraldez

**Comment by kikito** [comment]
@vm-001 since we are going to go with option 2, I will close this ticket. Please reopen if needed.

**Comment by vm-001** [comment]
I saw you approve the EE PR, which is option 1. Are taking option 1?
